File: course_4/module_2/programming_assignment_2/solution.py
import math


class TravelingSalesmanProblem:
    """
    A class to represent and solve the Traveling Salesman Problem using recursion and memoization.
    """

    def __init__(self, filename):
        """
        Initialize the TSP solver by reading data from the given file.
        :param filename: Path to the file containing the TSP instance.
        """
        self.filename = filename
        self.num_cities = 0
        self.coordinates = []
        self.distance_matrix = []
        self.VISITED_ALL = 0
        self.memo = {}

        # Read data from the file and prepare the distance matrix.
        self._load_data()
        self._prepare_distance_matrix()
        self.VISITED_ALL = (1 << self.num_cities) - 1  # Bitmask for all cities visited.

    def _load_data(self):
        """
        Load the number of cities and their coordinates from the file.
        """
        with open(self.filename, 'r') as file:
            lines = file.readlines()
            self.num_cities = int(lines[0].strip())  # First line specifies the number of cities.
            for line in lines[1:]:
                x, y = map(float, line.strip().split())
                self.coordinates.append((x, y))  # Store the coordinates as tuples (x, y).

    def _prepare_distance_matrix(self):
        """
        Prepare the distance matrix based on Euclidean distances between cities.
        """
        self.distance_matrix = [[0] * self.num_cities for _ in range(self.num_cities)]
        for i in range(self.num_cities):
            for j in range(self.num_cities):
                if i != j:
                    self.distance_matrix[i][j] = self._euclidean_distance(self.coordinates[i], self.coordinates[j])

    @staticmethod
    def _euclidean_distance(city1, city2):
        """
        Calculate the Euclidean distance between two cities.
        :param city1: Coordinates (x, y) of the first city.
        :param city2: Coordinates (x, y) of the second city.
        :return: Euclidean distance between the two cities.
        """
        x1, y1 = city1
        x2, y2 = city2
        return math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)

    def _solve_recursive(self, mask, pos):
        """
        Recursive function to compute the minimum cost of the TSP using bitmasking and memoization.
        :param mask: Bitmask representing visited cities.
        :param pos: Current city position.
        :return: Minimum cost of visiting all cities and returning to the start.
        """
        # Base case: all cities have been visited.
        if mask == self.VISITED_ALL:
            return self.distance_matrix[pos][0]  # Return the cost to return to the starting city.

        # Check if the result for this state is already memoized.
        key = (mask, pos)
        if key in self.memo:
            return self.memo[key]

        # Try visiting all unvisited cities and find the minimum cost.
        ans = math.inf
        for city in range(self.num_cities):
            if not (mask & (1 << city)):  # If the city has not been visited.
                new_cost = self.distance_matrix[pos][city] + self._solve_recursive(mask | (1 << city), city)
                ans = min(ans, new_cost)

        # Memoize the result for this state.
        self.memo[key] = ans
        return ans

    def solve(self):
        """
        Solve the TSP problem.
        :return: The minimum cost of the traveling salesman tour.
        """
        # Start the recursion from the first city with only the first city visited (mask = 1).
        return self._solve_recursive(1, 0)


# Example usage:
if __name__ == "__main__":
    # Path to the TSP instance file.
    filename = "tsp.txt"

    # Initialize the TSP solver with the input file.
    tsp_solver = TravelingSalesmanProblem(filename)

    # Solve the TSP and print the minimum cost.
    min_cost = tsp_solver.solve()
    print(f"Minimum cost: {math.floor(min_cost)}")

File: course_4/module_2/programming_assignment_2/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

// TravelingSalesmanProblem represents a solver for the TSP using recursion and memoization.
type TravelingSalesmanProblem struct {
	numCities      int                // Number of cities.
	coordinates    [][]float64        // Coordinates of the cities.
	distanceMatrix [][]float64        // Distance matrix between cities.
	VISITED_ALL    int                // Bitmask representing all cities visited.
	memo           map[string]float64 // Memoization table for subproblem results.
}

// NewTravelingSalesmanProblem initializes a new TSP solver by reading data from a file.
func NewTravelingSalesmanProblem(filename string) (*TravelingSalesmanProblem, error) {
	tsp := &TravelingSalesmanProblem{
		coordinates:    [][]float64{},
		distanceMatrix: [][]float64{},
		memo:           make(map[string]float64),
	}

	// Load data from the file.
	if err := tsp.loadData(filename); err != nil {
		return nil, err
	}

	// Prepare the distance matrix and the bitmask for all cities visited.
	tsp.prepareDistanceMatrix()
	tsp.VISITED_ALL = (1 << tsp.numCities) - 1

	return tsp, nil
}

// loadData reads the number of cities and their coordinates from the input file.
func (tsp *TravelingSalesmanProblem) loadData(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("error opening file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// Read the number of cities.
	if !scanner.Scan() {
		return fmt.Errorf("file is empty or invalid")
	}
	tsp.numCities, err = strconv.Atoi(scanner.Text())
	if err != nil {
		return fmt.Errorf("invalid number of cities: %v", err)
	}

	// Read the coordinates of the cities.
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) != 2 {
			return fmt.Errorf("invalid format for coordinates: %v", line)
		}

		x, err1 := strconv.ParseFloat(parts[0], 64)
		y, err2 := strconv.ParseFloat(parts[1], 64)
		if err1 != nil || err2 != nil {
			return fmt.Errorf("invalid coordinates: %v", line)
		}
		tsp.coordinates = append(tsp.coordinates, []float64{x, y})
	}

	return scanner.Err()
}

// prepareDistanceMatrix calculates the distance matrix using Euclidean distances.
func (tsp *TravelingSalesmanProblem) prepareDistanceMatrix() {
	tsp.distanceMatrix = make([][]float64, tsp.numCities)
	for i := 0; i < tsp.numCities; i++ {
		tsp.distanceMatrix[i] = make([]float64, tsp.numCities)
		for j := 0; j < tsp.numCities; j++ {
			if i != j {
				tsp.distanceMatrix[i][j] = tsp.euclideanDistance(tsp.coordinates[i], tsp.coordinates[j])
			}
		}
	}
}

// euclideanDistance calculates the Euclidean distance between two points.
func (tsp *TravelingSalesmanProblem) euclideanDistance(city1, city2 []float64) float64 {
	x1, y1 := city1[0], city1[1]
	x2, y2 := city2[0], city2[1]
	return math.Sqrt((x2-x1)*(x2-x1) + (y2-y1)*(y2-y1))
}

// solveRecursive is the core recursive function to solve the TSP.
func (tsp *TravelingSalesmanProblem) solveRecursive(mask, pos int) float64 {
	// Base case: all cities have been visited.
	if mask == tsp.VISITED_ALL {
		return tsp.distanceMatrix[pos][0] // Return to the starting city.
	}

	// Check memoization table.
	key := fmt.Sprintf("%d|%d", mask, pos)
	if val, exists := tsp.memo[key]; exists {
		return val
	}

	// Try all unvisited cities and find the minimum cost.
	ans := math.Inf(1)
	for city := 0; city < tsp.numCities; city++ {
		if mask&(1<<city) == 0 { // If the city has not been visited.
			newCost := tsp.distanceMatrix[pos][city] + tsp.solveRecursive(mask|(1<<city), city)
			if newCost < ans {
				ans = newCost
			}
		}
	}

	// Store the result in the memoization table.
	tsp.memo[key] = ans
	return ans
}

// Solve computes the minimum cost of the traveling salesman tour.
func (tsp *TravelingSalesmanProblem) Solve() float64 {
	// Start with the first city visited and position at the first city.
	return tsp.solveRecursive(1, 0)
}

func main() {
	// Path to the TSP instance file.
	filename := "course_4/module_2/programming_assignment_2/tsp.txt"

	// Initialize the TSP solver.
	tsp, err := NewTravelingSalesmanProblem(filename)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}

	// Solve the TSP and print the minimum cost.
	minCost := tsp.Solve()
	fmt.Printf("Minimum cost: %.0f\n", math.Floor(minCost))
}
